Allow whitelisting using __pdoc__

jpoeppel opened the pull request:
This PR implements whitelisting functionality for pdoc3 (#90). If the `__pdoc__` dictionary on a module level now contains a function or class name with the boolean value True, this function/class will now also be documented, even if it is otherwise marked by as private, i.e. by starting with an '_'. Also added a testcase for this behavior.

Since I just took a quick shot at it because I run into the case that I wanted to document a `__call__` override for one of my classes, I am not sure if I missed something. 

So far I assume that the whitelisting will always be done on the module level, i.e. the `__pdoc__` should be defined there instead of inside the class/function etc and just contain the full name, e.g. `A.__call__`. I think keeping all whitelisting together in one place is better than having them spread across all classes of a module, but let me know if you disagree.

One thing that I did not add, was "private" Parameters of functions, primarily since this would have required me to change the `_params` static function in order to get the function's name there, but also because I feel like specifying these parameters may be tedious.

Happy to get your comments on this.

kernc commented on pdoc/test/__init__.py:
```diff
@@ -513,6 +513,28 @@ def test__pdoc__invalid_value(self):
+            pdoc.link_inheritance()
+            self.assertIn('_private_function', mod.doc)
+
+
```
Please add two more tests:
```py
# Defined in example_pkg, referring to a member of its submodule 
with patch.object(module, '__pdoc__', {'subpkg.A.__call__': True}):
```
and
```py
# Using full refname
with patch.object(module, '__pdoc__', {'example_pkg.subpkg.A.__call__': True}):
```

jpoeppel replied:
I am currently struggeling with these two tests, as I do not know what the best way would be to pass down the __pdoc__ of example_pkg down to subpkg. So far from my tests, it seems like subpkg will have it's own __pdoc__. I can check for supermodules and look there, but this would result in an ugly mess of special checks. The cleaner solution would probably be to update a subpackages/submodule's __pdoc__ with the supermodule's __pdoc__ or at least with the relevant elements. Am I missing some way of doing this already?

kernc replied:
Just as a quick idea, before I even view the code, we have `pdoc.Context`.
https://github.com/pdoc3/pdoc/blob/196c93bbf0964f971ab0e2f323f9fac29fe7a7f8/pdoc/__init__.py#L63-L75
Maybe something can be saved in there for the children modules in advance?

kernc replied:
`pdoc.Context` certainly is how the core `pdoc.Module.find_ident()` works. I think this might be a clearer approach ...

jpoeppel replied:
Do you suggest filling Module._context with the conents of __pdoc__ or rather inserting the dictionary __pdoc__ in the context so that it is available for submodules? 
The first option could lead to unintended behavior, as find_ident is used to get a Doc-Object but __pdoc__ would be having booleans or Strings as values which may even overwrite existing elements.
The second would make the assumption that there should never be anything named "__pdoc__", which might be ok.

kernc replied:
Context should certainly remain mapping only refnames to doc objects. As a third option, we could attach an extra property on it, like `Context.__pdoc__`, a conjunction of all `__pdoc__` dicts we came across so far.

jpoeppel replied:
Not sure if this would be cleaner, as we then run into the problem discussed below with relative vs absolute references. You would need to either filter out references for the current module, or at least modify them so that submodules can clearly tell that those names are not relevant for them.

kernc commented on pdoc/test/example_pkg/__init__.py:
```diff
@@ -24,6 +24,11 @@ def object_as_arg_default(*args, a=object(), **kwargs):
+    """Private function, should only appear if whitelisted"""
+
+
+
```
[CI is failing](https://travis-ci.org/pdoc3/pdoc/jobs/635340060#L241-L263) due to this extra whitespace.

jpoeppel replied:
Should be fixed now.

jpoeppel commented on pdoc/__init__.py:
```diff
@@ -274,6 +274,15 @@ def _pep224_docstrings(doc_obj: Union['Module', 'Class'], *,
+    Returns `True` if `ident_name` is contained in the module's __pdoc__ 
+    with a value of `True`.
+    """
+    if module:
```
You are right, this is not necessary anymore. When I was still considering allowing it for params as well, it may be None as the _params method has the module only as optional parameter.

jpoeppel commented on pdoc/__init__.py:
```diff
@@ -650,6 +661,8 @@ def _link_inheritance(self):
                         del self._context[key]
 
                 continue
+            if docstring is True:
```
I inserted it to prevent the ValueError that assumes it to be a string at this point, but I've added a comment describing this.

kernc commented on pdoc/test/__init__.py:
```diff
@@ -513,6 +513,48 @@ def test__pdoc__invalid_value(self):
+            self.assertIn('subpkg', mod.doc)
+            self.assertIn('A', mod.doc['subpkg'].doc)
+            self.assertIn('__call__', mod.doc['subpkg'].doc['A'].doc)
+
```
Let's also test whitelisting modules:
```py
with patch.object(module, '__pdoc__', {'example_pkg._private': True}):
    ...
with patch.object(module, '__pdoc__', {'_private': True}):
    ...
```

jpoeppel replied:
Those tests worked out of the box with the current implementation 👍

kernc commented:
Hmmm, I assumed docstrings can be overridden using canonical refnames in parent modules. Looking at the code, this appears to not be the case, only _current_ `__pdoc__` is iterated through. :thinking: 
https://github.com/pdoc3/pdoc/blob/196c93bbf0964f971ab0e2f323f9fac29fe7a7f8/pdoc/__init__.py#L627-L660

jpoeppel commented:
I guess the amiguousity is only really a problem, if a subpackage has the same name as its containing package and both packages contain an object (class or function) with the same name. In that case, a specification without the global refname could mean either (both for whitelisting and overwriting), right? 
Otherwise, you would know whether or not the string represents a relative or global reference. 

In terms of current behavior for the whitelisting implementation:
In case we have: 
package
  - _A
  - package
    - _A

and we whitelist using the relative reference`package._A: True` on the top level, we would whitelist both "_A" classes/functions, however if we were to use the global reference name `package.package._A: True` on top level, it should work as intended to only whitelist the lower one. However, if we only wanted to whitelist the top level `_A`, we would need to use the relative name `_A: True`.

kernc commented:
But, then, if we used just `_A: True`, that would whitelist any `_A` in any later module ...

jpoeppel commented:
> But, then, if we used just `_A: True`, that would whitelist any `_A` in any later module ...

This is not true with the current implementation, as submodules would look for "submodule._A" in the __pdoc__ of its supermodule. I've added a test for that with the _private_function.

kernc commented on pdoc/__init__.py:
```diff
@@ -275,6 +275,28 @@ def _pep224_docstrings(doc_obj: Union['Module', 'Class'], *,
     return vars, instance_vars
 
 
+def _is_whitelisted(ident_name, doc_obj):
```
`doc_obj` seems only used for `doc_obj.module`, so why not accept simply `doc_module` where whitelisted `ident_name` is sought in?

jpoeppel replied:
You are right, I changed this when I was figuring out a way to handle the absolute paths, but ended up only needing the module after all, changed this back in newest commit.

kernc replied:
Tidying up afterwards is part of the deal. :stuck_out_tongue:

jpoeppel replied:
True, thanks for spotting!

kernc commented on pdoc/__init__.py:
```diff
@@ -275,6 +275,28 @@ def _pep224_docstrings(doc_obj: Union['Module', 'Class'], *,
+
+    # We also need to check all potential supermodules' __pdoc__s
+    # due to full reference
+    if hasattr(doc_obj.module, "supermodule") and doc_obj.module.supermodule:
```
`pdoc.Module.supermodule` is always set?
https://github.com/pdoc3/pdoc/blob/72bd9821e6a8180230f220a453dd4c0d53832063/pdoc/__init__.py#L522-L525

jpoeppel replied:
This was also introduced when I was considering different kinds of objects (I was initially checking if "supermodule" was in doc_obj directly, but this has also been reverted. I still need the if check, because it may be None as far as I can tell.

jpoeppel commented:
I did add a test, using a string instead of "True" in the overwrite. The test checks if the object is whitelisted, however I am not sure how to test, if the overwrite works correctly, with regard to the resulting documentation. It should, as this should happen as before after the docs have been generated, all I am changing is the list of objects to create the doc for, all post processing should work as before.

kernc commented on pdoc/test/__init__.py:
```diff
@@ -513,6 +513,61 @@ def test__pdoc__invalid_value(self):
+            pdoc.link_inheritance()
+            self.assertIn('A', mod.doc)
+            self.assertIn('__call__', mod.doc['A'].doc)
+
```
I think `mod.doc['A'].doc['__call__'].docstring` should be the expected string.

jpoeppel replied:
Looks good, test passes when I check for that.

jpoeppel commented:
Any idea why Travis is listed as pending? When I check details it shows up as passed.

jpoeppel commented:
Is there anything still missing for this to be merged? The "problem" with relative vs global references is a general one I suppose, not something limited to this PR or am I missunderstanding something?

kernc commented:
The concern that popped up while fixing https://github.com/pdoc3/pdoc/commit/e8821eee42de2eaa4b56740a4ff1350f3902b354 was whether whitelisting also works for [instance variables](https://pdoc3.github.io/pdoc/doc/pdoc/#docstrings-for-variables).

That's not what's been stalling the PR, though. I should certainly get to in a few day. Thanks. :sweat_smile:

jpoeppel commented:
Good to know. I rebased and added a test for that. It obviously did not work right from the start, but when I check for the whitelist in the pep224 function, it does pass the tests. 
I do however need to do it a little uglier than at other places, because I need to check if doc_obj is a class or a module and prepaid the class' name in that case, because the whitelist would need to be for B._private_instance_var in order to be unique. 
Let me know if you do not like this approach and I revert that change.

kernc commented:
Good thing I take so long with this one. :sweat_smile: I noticed the documentation mention of the new/extended feature is still missing ... Understandably, I reserve the right to a final edit, but maybe you'd like to give copy a shot?

jpoeppel commented:
I updated the documentation.md file in the section regarding the overriding. Let me know what you think.

kernc commented:
Sorry I took so long with this one. It's an important feature, making the time well invested. Thanks for your work on this!
